File: src/domain/observation/freshness.ts
/**
 * Âge d'une observation hydrométrique temps réel (BR-005).
 *
 * Union close, sans valeur par défaut implicite (BR-011).
 */
export type Freshness = "Fraiche" | "Ancienne" | "Perimee";

/**
 * Seuils ABSOLUS fixés par BR-005 : 2 h puis 24 h.
 *
 * ⚠️ Ce ne sont pas des multiples du TTL de cache. `03-conception.md § 4.1`
 * évoque « 2 × TTL », mais il y décrit la fraîcheur du CACHE — une autre
 * question. Les confondre déclarerait périmée une observation de 40 minutes,
 * alors que BR-005 lui laisse 24 heures.
 */
export const ANCIENNE_APRES_MS = 2 * 60 * 60 * 1000;
export const PERIMEE_APRES_MS = 24 * 60 * 60 * 1000;

/**
 * Fonction pure : l'instant courant est un paramètre, jamais `Date.now()` lu à
 * l'intérieur. Ce sont les bornes qui portent la règle, et des bornes non
 * testables ne sont pas des bornes.
 *
 * `observedAt` est la date de MESURE (`date_obs`), jamais la date de
 * récupération — une donnée fraîchement téléchargée peut être vieille de neuf
 * jours, et c'est le cas en production (BR-001, BR-005).
 *
 * ⚠️ Ne s'applique pas aux observations ONDE : une campagne de trois semaines
 * y est normale, pas périmée. Voir BR-010.
 */
export function freshnessOf(observedAt: Date, now: Date): Freshness {
  const ageMs = now.getTime() - observedAt.getTime();

  // La borne appartient à l'état le plus sévère : on ne minimise jamais l'âge
  // d'une donnée sur laquelle un usager peut fonder une décision.
  if (ageMs >= PERIMEE_APRES_MS) return "Perimee";
  if (ageMs >= ANCIENNE_APRES_MS) return "Ancienne";
  return "Fraiche";
}

File: lib/domain/observation/freshness.dart
// Seuils absolus de BR-005, sans rapport avec un TTL de cache : les
// confondre déclarerait périmée une observation de quarante minutes.
// Ne s'applique pas aux observations d'écoulement, où une campagne de trois
// semaines est normale (BR-010).

/// État de fraîcheur d'une observation hydrométrique, calculé sur la date de
/// **mesure** — jamais de récupération (`BR-001`) : une donnée fraîchement
/// téléchargée peut avoir dix-sept jours.
///
/// Affichage attendu :
/// - [fraiche] : aucune mention à l'écran ;
/// - [ancienne] : « il y a N h » ;
/// - [perimee] : marqueur atténué **et** avertissement explicite.
enum Freshness {
  /// Âge strictement inférieur à [ancienneApres].
  fraiche,

  /// Âge dans `[ancienneApres, perimeeApres)`.
  ancienne,

  /// Âge supérieur ou égal à [perimeeApres].
  perimee,
}

/// Borne au-delà de laquelle une observation cesse d'être [Freshness.fraiche].
/// La borne appartient à l'état le plus sévère : un âge de 2 h exactement est
/// [Freshness.ancienne].
const Duration ancienneApres = Duration(hours: 2);

/// Borne au-delà de laquelle une observation devient [Freshness.perimee].
/// La borne appartient à l'état le plus sévère : un âge de 24 h exactement
/// est [Freshness.perimee].
const Duration perimeeApres = Duration(hours: 24);

/// Calcule la fraîcheur d'une observation mesurée à [measuredAt], vue à
/// l'instant [now]. Fonction pure : [now] est un paramètre, jamais lu à
/// l'intérieur — c'est ce qui rend les bornes testables.
///
/// Un âge négatif (mesure dans le futur) est une anomalie de la source, pas
/// une donnée vieille : il est traité comme [Freshness.fraiche] plutôt que
/// d'inventer un âge inconnu.
///
/// `now.difference(measuredAt)` compare des instants absolus : la
/// comparaison reste correcte que l'une ou l'autre des deux dates soit
/// exprimée en UTC ou en heure locale, sans conversion préalable.
Freshness freshnessOf({required DateTime measuredAt, required DateTime now}) {
  final Duration age = now.difference(measuredAt);

  if (age < ancienneApres) {
    return Freshness.fraiche;
  }
  if (age < perimeeApres) {
    return Freshness.ancienne;
  }
  return Freshness.perimee;
}
